cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package throttling
+
+import (
+	"math"
+	"time"
+)
+
+type tokenBucket struct {
+	creditsPerSecond float64
+	balance          float64
+	maxBalance       float64
+	lastTick         time.Time
+	timeNow          func() time.Time // For testing
+}
+
+func (t *tokenBucket) Drain(maxDrain float64) float64 {
+	now := t.timeNow()
+	interval := now.Sub(t.lastTick)
+	t.lastTick = now
+	diff := interval.Seconds() * t.creditsPerSecond
+	t.balance = math.Min(t.balance+diff, t.maxBalance)
+	result := math.Min(t.balance, maxDrain)
+	t.balance -= result
+	return result
+}
+
+type client struct {
+	id                  string
+	perOperationBalance map[string]float64
+	defaultBalance      float64
+	updateTime          time.Time
+}
+
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance, ok := c.perOperationBalance[operationName]
+	if !ok {
+		balance = c.defaultBalance
+	}
+
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	if ok {
+		c.perOperationBalance[operationName] = balance
+	} else {
+		c.defaultBalance = balance
+	}
+	return true
+}
+
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
+}
+
+type creditAccruer struct {
+	options                 CreditAccruerOptions
+	serviceName             string
+	perOperationRateLimiter map[string]*tokenBucket
+	defaultRateLimiter      *tokenBucket
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func (ca *creditAccruer) Withdraw(operationName string, balance float64) float64 {
+	if balance >= ca.options.ClientMaxBalance {
+		return 0
+	}
+	rateLimiter, ok := ca.perOperationRateLimiter[operationName]
+	if !ok {
+		if len(ca.perOperationRateLimiter) < ca.options.MaxOperations {
+			rateLimiter = &tokenBucket{
+				creditsPerSecond: ca.options.CreditsPerSecond,
+				balance:          ca.options.MaxBalance,
+				maxBalance:       ca.options.MaxBalance,
+				timeNow:          ca.timeNow,
+				lastTick:         ca.timeNow(),
+			}
+			ca.perOperationRateLimiter[operationName] = rateLimiter
+		} else {
+			rateLimiter = ca.defaultRateLimiter
+		}
+	}
+
+	credits := rateLimiter.Drain(ca.options.ClientMaxBalance - balance)
+	return credits
+}
+
+type creditStore struct {
```

Comment:
documentation. I don't follow the business meaning of this struct and the two maps it keeps.

Reply:
Will do ASAP